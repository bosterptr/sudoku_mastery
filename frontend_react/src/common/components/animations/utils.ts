import { CSSProperties } from 'react';

export const reflow = (node: Element) => node.scrollTop;

interface ComponentProps {
  easing: string | { enter?: string; exit?: string } | undefined;
  style: CSSProperties | undefined;
  timeout: number | { enter?: number; exit?: number };
}

interface Options {
  mode: 'enter' | 'exit';
}

interface TransitionProps {
  duration: string | number;
  easing: string | undefined;
  delay: string | undefined;
}

export function getTransitionProps(props: ComponentProps, options: Options): TransitionProps {
  const { timeout, easing, style = {} } = props;

  return {
    duration:
      style.transitionDuration ??
      (typeof timeout === 'number' ? timeout : timeout[options.mode] || 0),
    easing:
      style.transitionTimingFunction ??
      (typeof easing === 'object' ? easing[options.mode] : easing),
    delay: style.transitionDelay,
  };
}

export function getAutoHeightDuration(height: number) {
  if (!height) return 0;
  const constant = height / 36;
  // https://www.wolframalpha.com/input/?i=(4+%2B+15+*+(x+%2F+36+)+**+0.25+%2B+(x+%2F+36)+%2F+5)+*+10
  return Math.round((4 + 15 * constant ** 0.25 + constant / 5) * 10);
}
