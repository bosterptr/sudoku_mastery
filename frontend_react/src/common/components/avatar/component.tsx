import { CSSProperties, ReactNode } from 'react';
import styled, { css } from 'styled-components';

const Image = styled.img`
  border-radius: 50%;
  display: flex;
  height: 100%;
  position: relative;
  width: 100%;
`;

const small = css`
  height: 30px;
  width: 30px;
`;
const medium = css`
  height: 40px;
  width: 40px;
`;
const large = css`
  height: 48px;
  width: 48px;
`;
const profile = css`
  height: 180px;
  width: 180px;
`;
const smallText = css`
  font-size: 12px;
`;
const mediumText = css`
  font-size: 18px;
`;
const largeText = css`
  font-size: 26px;
`;
const profileText = css`
  font-size: 80px;
`;
// eslint-disable-next-line no-shadow
export const AvatarVariant = {
  CIRCLE: 0,
  ROUNDED: 1,
  SQUARE: 2,
} as const;

type IAvatarVariantValue = (typeof AvatarVariant)[keyof typeof AvatarVariant];
const Container = styled.div<{
  size: AvatarSizeType;
  $variant?: IAvatarVariantValue;
}>`
  & > * {
    font-weight: 500;
    ${(p) => p.size === 'large' && largeText}
    ${(p) => p.size === 'medium' && mediumText}
    ${(p) => p.size === 'small' && smallText}
    ${(p) => p.size === 'profile' && profileText}
    ${(p) =>
      typeof p.size === 'number' &&
      css`
        font-size: ${Math.floor(p.size * 0.45)}px;
      `}
  }
  ${(p) => p.size === 'large' && large}
  ${(p) => p.size === 'medium' && medium}
  ${(p) => p.size === 'small' && small}
  ${(p) => p.size === 'profile' && profile}
  ${(p) =>
    typeof p.size === 'number' &&
    css`
      height: ${p.size}px;
      width: ${p.size}px;
    `}
  align-items: center;
  background-color: rgb(0 87 242);
  border-radius: ${(p) => {
    switch (p.$variant) {
      case AvatarVariant.ROUNDED:
        return '6px';
      case AvatarVariant.SQUARE:
        return '0px';
      default:
        return '50%';
    }
  }};
  border: 2px solid #fff;
  color: white;
  display: flex;
  justify-content: center;
  overflow: hidden;
  position: relative;
`;

export type AvatarSizeType = 'small' | 'medium' | 'large' | 'profile' | number;

interface Props {
  alt?: string;
  src?: string | null;
  style?: CSSProperties;
  onClick?: () => void;
  // eslint-disable-next-line react/require-default-props
  size?: AvatarSizeType;
  className?: string;
  children?: ReactNode;
  variant?: IAvatarVariantValue;
}
export const Avatar = ({
  style,
  src,
  onClick,
  size = 'medium',
  className,
  children,
  alt,
  variant = AvatarVariant.CIRCLE,
}: Props) => {
  return (
    <Container className={className} onClick={onClick} size={size} style={style} $variant={variant}>
      {src && <Image alt={alt} src={src} />}
      {!src && <span>{children}</span>}
    </Container>
  );
};
