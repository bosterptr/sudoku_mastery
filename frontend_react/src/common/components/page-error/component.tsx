import { ReactNode } from 'react';
import styled, { css } from 'styled-components';

const Root = styled.div<{ $fullWidth: boolean; $fullHeight: boolean }>`
  display: flex;
  color: ${(p) => p.theme.palette.text.secondary};
  justify-content: center;
  align-items: center;
  ${(p) =>
    p.$fullWidth &&
    css`
      width: 100%;
    `}
  ${(p) =>
    p.$fullHeight &&
    css`
      height: 100%;
    `}
`;

const Container = styled.div`
  display: inline-flex;
  align-items: center;
`;

const SubContainer = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  height: 100%;
`;

const CenteredText = styled.div`
  font-size: 16px;
`;

export interface Props {
  children?: ReactNode;
  inline?: boolean;
  message?: string;
}

export const PageError = ({ children, inline = false, message }: Props) => (
  <Root $fullWidth={!inline} $fullHeight={!inline}>
    <Container>
      <SubContainer>
        <CenteredText>{message}</CenteredText>
        {children}
      </SubContainer>
    </Container>
  </Root>
);
