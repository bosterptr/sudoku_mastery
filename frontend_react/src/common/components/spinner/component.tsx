import * as duration from 'app/core/themes/duration';
import { ComponentSize } from 'app/core/themes/types';
import styled, { css, keyframes } from 'styled-components';

export interface LoadingSpinnerProps {
  size?: ComponentSize;
  /** Delay the appearance of the loading spinner; if you unmount the component before this time has elapsed, the user will not see a loading spinner */
  delay?: number;
  inheritColor?: boolean;
  fillContent?: boolean;
}

const loadingSpinnerAnimation = keyframes`
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(360deg);
  }
`;

const loadingSpinnerAppear = keyframes`
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
`;

const SpinnerContainer = styled.div<{ delay?: number; $fillContent?: boolean }>`
  ${({ $fillContent }) =>
    $fillContent &&
    css`
      align-items: center;
      display: flex;
      height: 100%;
      justify-content: center;
      width: 100%;
    `}

  ${({ delay }) =>
    Boolean(delay) &&
    css`
      animation-delay: ${delay}ms;
      animation-duration: ${duration.standard};
      animation-fill-mode: forwards;
      animation-iteration-count: 1;
      animation-name: ${loadingSpinnerAppear};
      animation-timing-function: ease;
      opacity: 0;
    `}
`;
function getPropertiesForSize(size?: ComponentSize) {
  switch (size) {
    case 'xs':
      return {
        width: '1.2rem',
        height: '1.2rem',
      };
    case 'sm':
      return {
        width: '1.6rem',
        height: '1.6rem',
      };
    case 'lg':
      return {
        width: '2.8rem',
        height: '2.8rem',
      };
    case 'md':
    default:
      return {
        width: '2.2rem',
        height: '2.2rem',
      };
  }
}

const SpinnerCircle = styled.div<{ $inheritColor?: boolean; size?: ComponentSize }>`
  border-right: 4px solid transparent;
  border-top: 4px solid #d1d1d1;
  border-radius: 50%;
  position: relative;
  transform: translateZ(0);
  animation-name: ${loadingSpinnerAnimation};
  animation-iteration-count: infinite;
  animation-duration: 1s;
  animation-timing-function: linear;
  ${(p) => getPropertiesForSize(p.size)}
  ${({ $inheritColor }) =>
    $inheritColor &&
    css`
      border-top-color: ${(p) => p.theme.palette.border.weak};
      border-right-color: ${(p) => p.theme.palette.border.weak};
      border-bottom-color: ${(p) => p.theme.palette.border.weak};
      border-left-color: currentColor;
    `}
`;

export const LoadingSpinner = ({
  delay = 300,
  fillContent,
  inheritColor,
  size,
}: LoadingSpinnerProps) => (
  <SpinnerContainer delay={delay} $fillContent={fillContent} data-testid="spinner">
    <SpinnerCircle $inheritColor={inheritColor} size={size} />
  </SpinnerContainer>
);
