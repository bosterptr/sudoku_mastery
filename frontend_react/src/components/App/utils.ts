const GL_BUFFER_STATUS_CHECK_INTERVAL = 2;

export type IOpenglSetup = [
  OffscreenCanvas,
  WebGLProgram,
  WebGLProgram,
  WebGLProgram,
  ITexFbPair,
  ITexFbPair,
  number,
];

export interface ITexFbPair {
  tex: WebGLTexture;
  fb: WebGLFramebuffer;
}
export const getGrayScaleFromUint8Array = (
  dataArray: Uint8Array,
  width: number,
  height: number,
  cachedGrayScaleUintArray: Uint8Array,
  cachedGrayScaleSize: number,
): Uint8Array => {
  if (cachedGrayScaleSize !== width * height) {
    // eslint-disable-next-line no-param-reassign
    cachedGrayScaleUintArray = new Uint8Array(width * height);
    // eslint-disable-next-line no-param-reassign
    cachedGrayScaleSize = width * height;
  }
  const { length } = dataArray;
  for (let i = 0, luminanceIndex = 0; i < length; i += 4, luminanceIndex += 1) {
    // eslint-disable-next-line no-bitwise
    const r = dataArray[i] >> 2;
    // eslint-disable-next-line no-bitwise
    const g = dataArray[i + 1] >> 1;
    // eslint-disable-next-line no-bitwise
    const b = dataArray[i + 2] >> 3;
    // const luminance = 0.299 * r + 0.587 * g + 0.114 * b;
    // luminanceData[luminanceIndex] =
    //   0.299 * dataArray[i] +
    //   0.587 * dataArray[i + 1] +
    //   0.114 * dataArray[i + 2];
    // eslint-disable-next-line no-param-reassign
    cachedGrayScaleUintArray[luminanceIndex] = r + b + g;
  }
  return cachedGrayScaleUintArray;
};

export const clientWaitAsync = (
  gl: WebGL2RenderingContext,
  sync: WebGLSync,
  flags = 0,
  intervalMs = 10,
) => {
  return new Promise<void>((resolve, reject) => {
    const check = () => {
      const res = gl.clientWaitSync(sync, flags, 0);
      if (res === gl.WAIT_FAILED) {
        reject();
        return;
      }
      if (res === gl.TIMEOUT_EXPIRED) {
        setTimeout(check, intervalMs);
        return;
      }
      resolve();
    };
    check();
  });
};

export const readPixelsAsync = (
  gl: WebGL2RenderingContext,
  width: number,
  height: number,
  buffer: Uint8Array,
): Promise<Uint8Array> => {
  const bufpak = gl.createBuffer();
  gl.bindBuffer(gl.PIXEL_PACK_BUFFER, bufpak);
  gl.bufferData(gl.PIXEL_PACK_BUFFER, buffer.byteLength, gl.STREAM_READ);
  gl.readPixels(0, 0, width, height, gl.RGBA, gl.UNSIGNED_BYTE, 0);
  const sync = gl.fenceSync(gl.SYNC_GPU_COMMANDS_COMPLETE, 0);
  if (!sync) throw new Error('WebGLSync is empty');
  gl.flush();
  return clientWaitAsync(gl, sync, 0, GL_BUFFER_STATUS_CHECK_INTERVAL).then(() => {
    gl.deleteSync(sync);
    gl.bindBuffer(gl.PIXEL_PACK_BUFFER, bufpak);
    gl.getBufferSubData(gl.PIXEL_PACK_BUFFER, 0, buffer);
    gl.bindBuffer(gl.PIXEL_PACK_BUFFER, null);
    gl.deleteBuffer(bufpak);
    return buffer;
  });
};

export function setAttributes(gl: WebGL2RenderingContext, program: WebGLProgram, name: string) {
  const location = gl.getAttribLocation(program, name);
  gl.enableVertexAttribArray(location);
  gl.vertexAttribPointer(location, 2, gl.FLOAT, false, 0, 0);
}

export function isMobile() {
  const userAgent = navigator.userAgent.toLowerCase();
  return /iphone|ipod|android|ie|blackberry|fennec/.test(userAgent);
}
export const getCameraResolution = () => {
  if (isMobile()) {
    return {
      width: { ideal: document.body.clientWidth > 500 ? 500 : document.body.clientWidth },
      height: { ideal: document.body.clientHeight > 500 ? 500 : document.body.clientHeight },
    };
  }
  return {
    width: { ideal: 1000 },
    height: { ideal: 1000 },
  };
};
