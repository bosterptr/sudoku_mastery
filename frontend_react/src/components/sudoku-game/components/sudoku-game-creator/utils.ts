import { ALL_SUDOKU_CELLS, CANVAS_SIZE, SUDOKU_CELLS_IN_ROW } from 'app/core/constants/sudoku';
import { ICellNotes, IMission } from './types';

export const generateNotes = (): ICellNotes[] => {
  const newArray: ICellNotes[] = [];
  for (let cellIndex = 0; cellIndex < ALL_SUDOKU_CELLS; cellIndex += 1) {
    newArray.push([]);
  }
  return newArray;
};
export const getBlockSize = (pixelRatio: number) => (CANVAS_SIZE * pixelRatio - 2 * pixelRatio) / 3;
export const getAdjustedPixelRatio = () => {
  const { devicePixelRatio } = window;
  return devicePixelRatio > 2 ? 2 * devicePixelRatio : 4;
};
export const getCellSize = (pixelRatio: number) => (CANVAS_SIZE * pixelRatio - 4 * pixelRatio) / 9;
export const getContextSize = (pixelRatio: number) => CANVAS_SIZE * pixelRatio;
export const getOuterSize = (pixelRatio: number) => 2 * pixelRatio;
export const getThinSize = (pixelRatio: number) => 0.75 * pixelRatio;
export const getX = (index: number) => Math.floor(index / 9);
export const getY = (index: number) => index % 9;
export const getCellXPosition = (outerSize: number, cellSize: number, index: number) =>
  outerSize + cellSize * getY(index);
export const getCellYPosition = (outerSize: number, cellSize: number, index: number) =>
  outerSize + cellSize * getX(index);
export const getCubeId = (index: number) =>
  `${Math.floor(getX(index) / 3)}${Math.floor(getY(index) / 3)}`;
export const EMPTY_CELL = '.';
export const isSudokuCellEmpty = (char: string) => char === EMPTY_CELL;
export const generateArrWithTrueAtIndexes = (indexes: number[]): boolean[] => {
  const newArray: boolean[] = [];
  for (let cellIndex = 0; cellIndex < ALL_SUDOKU_CELLS; cellIndex += 1) {
    if (indexes.includes(cellIndex)) newArray[cellIndex] = true;
    else newArray[cellIndex] = false;
  }
  return newArray;
};
export const getScoreFromMove = (values: IMission, solution: string, selectedCell: number) => {
  let fullRows = 0;
  let fullColumns = 0;
  let fullCubes = 0;
  for (let cellIndex = 0; cellIndex < ALL_SUDOKU_CELLS; cellIndex += 1) {
    if (values[cellIndex] === solution[cellIndex]) {
      if (getY(cellIndex) === getY(selectedCell)) fullRows += 1;
      if (getX(cellIndex) === getX(selectedCell)) fullColumns += 1;
      if (getCubeId(cellIndex) === getCubeId(selectedCell)) fullCubes += 1;
    }
  }
  let score = 1;
  if (fullRows === SUDOKU_CELLS_IN_ROW) score += 5;
  if (fullColumns === SUDOKU_CELLS_IN_ROW) score += 5;
  if (fullCubes === 9) score += 5;
  return score;
};
