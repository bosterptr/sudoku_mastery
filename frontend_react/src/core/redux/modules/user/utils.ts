import { IUser } from 'app/core/redux/modules/user/types';

// eslint-disable-next-line import/prefer-default-export
export function normalizeSetUsers(users: Record<string, IUser>) {
  const normalizedUsers: Record<string, IUser> = {};
  Object.entries(users).forEach(([, user]) => {
    normalizedUsers[user.id] = user;
  });

  return normalizedUsers;
}
