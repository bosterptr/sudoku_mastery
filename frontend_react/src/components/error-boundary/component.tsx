/* eslint-disable react/destructuring-assignment */
import Button from 'app/common/components/button';
import { logger } from 'app/core/App';
import { Component, ErrorInfo, ReactNode } from 'react';
import { FormattedMessage } from 'react-intl';
import { ErrorImageContainer, ErrorImageOverlay, ErrorImageText } from './styles';

export interface Props {
  children?: ReactNode;
  name: string;
  content?: JSX.Element;
  onError?: () => void;
}

interface State {
  hasErrored: boolean;
}

export class ErrorBoundary extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { hasErrored: false };
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public componentDidCatch(error: Error, _info: ErrorInfo) {
    this.setState({ hasErrored: true });
    logger.errorAndReport(error, this.props.name);
    if (this.props.onError) {
      this.props.onError();
    }
  }

  public render() {
    if (this.state.hasErrored) {
      if (this.props.content) {
        return this.props.content;
      }
      return (
        <ErrorImageOverlay>
          <ErrorImageContainer $imageurl="https://i.imgur.com/yW2W9SC.png" />
          <ErrorImageText>
            <FormattedMessage defaultMessage="Sorry this page is broken" />
          </ErrorImageText>
          <Button aria-label="Reload page" onClick={() => window.location.reload()}>
            <FormattedMessage defaultMessage="Reload page" />
          </Button>
        </ErrorImageOverlay>
      );
    }
    // eslint-disable-next-line react/destructuring-assignment
    return this.props.children;
  }
}
