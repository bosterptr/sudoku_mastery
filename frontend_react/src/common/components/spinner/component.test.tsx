/* eslint-disable react/jsx-props-no-spreading */
import { render } from 'app/tests/utils';
import { LoadingSpinner, LoadingSpinnerProps } from './component';

function getOptionalProps(): LoadingSpinnerProps {
  return {
    size: 'lg',
    delay: 100,
    inheritColor: true,
    fillContent: true,
  };
}

describe('LoadingSpinner', () => {
  it('renders loading spinner component with required props', () => {
    const { asFragment } = render(<LoadingSpinner />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders loading spinner component with all props', () => {
    const { asFragment } = render(<LoadingSpinner {...getOptionalProps()} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('has a delay by default', () => {
    const { asFragment } = render(<LoadingSpinner />);
    expect(asFragment().firstChild).toHaveStyle('animation-delay: 300ms');
  });

  it('does not add a delay className with 0 delay', () => {
    const { asFragment } = render(<LoadingSpinner delay={0} />);
    expect(asFragment()).not.toHaveStyleRule('animation-delay');
  });
});
