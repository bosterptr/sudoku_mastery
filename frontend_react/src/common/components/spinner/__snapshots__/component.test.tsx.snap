// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingSpinner renders loading spinner component with all props 1`] = `
<DocumentFragment>
  .c0 {
  align-items: center;
  display: flex;
  height: 100%;
  justify-content: center;
  width: 100%;
  animation-delay: 100ms;
  animation-duration: 300;
  animation-fill-mode: forwards;
  animation-iteration-count: 1;
  animation-name: cJoqxJ;
  animation-timing-function: ease;
  opacity: 0;
}

.c1 {
  border-right: 4px solid transparent;
  border-top: 4px solid #d1d1d1;
  border-radius: 50%;
  position: relative;
  transform: translateZ(0);
  animation-name: epmXAj;
  animation-iteration-count: infinite;
  animation-duration: 1s;
  animation-timing-function: linear;
  width: 2.8rem;
  height: 2.8rem;
  border-top-color: #24292e1f;
  border-right-color: #24292e1f;
  border-bottom-color: #24292e1f;
  border-left-color: currentColor;
}

<div
    class="c0"
    data-testid="spinner"
    delay="100"
  >
    <div
      class="c1"
    />
  </div>
</DocumentFragment>
`;

exports[`LoadingSpinner renders loading spinner component with required props 1`] = `
<DocumentFragment>
  .c0 {
  animation-delay: 300ms;
  animation-duration: 300;
  animation-fill-mode: forwards;
  animation-iteration-count: 1;
  animation-name: cJoqxJ;
  animation-timing-function: ease;
  opacity: 0;
}

.c1 {
  border-right: 4px solid transparent;
  border-top: 4px solid #d1d1d1;
  border-radius: 50%;
  position: relative;
  transform: translateZ(0);
  animation-name: epmXAj;
  animation-iteration-count: infinite;
  animation-duration: 1s;
  animation-timing-function: linear;
  width: 2.2rem;
  height: 2.2rem;
}

<div
    class="c0"
    data-testid="spinner"
    delay="300"
  >
    <div
      class="c1"
    />
  </div>
</DocumentFragment>
`;
